python_modules/libraries/dagster-airbyte/dagster_airbyte/managed/types.py

```diff
@@ -0,0 +1,182 @@
+from enum import Enum
+from typing import Any, Dict, Mapping, Optional
+
+
+class AirbyteSyncMode(Enum):
+    """
+    Represents the sync mode for a given Airbyte stream.
+    """
+
+    FULL_REFRESH_APPEND = ("full_refresh", "append")
+    FULL_REFRESH_OVERWRITE = ("full_refresh", "overwrite")
+    INCREMENTAL_APPEND = ("incremental", "append")
+    INCREMENTAL_OVERWRITE = ("incremental", "overwrite")
+    INCREMENTAL_APPEND_DEDUP = ("incremental", "append_dedup")
+
+
+class AirbyteSource:
+    """
+    Represents a user-defined Airbyte source.
+    """
+
+    def __init__(self, name: str, source_type: str, source_configuration: Dict[str, Any]):
+        self.name = name
+        self.source_type = source_type
+        self.source_configuration = source_configuration
+
+    def must_be_recreated(self, other: "AirbyteSource") -> bool:
+        return self.name != other.name or self.source_configuration != other.source_configuration
+
+
+class InitializedAirbyteSource:
+    """
+    User-defined Airbyte source bound to actual created Airbyte source.
+    """
+
+    def __init__(self, source: AirbyteSource, source_id: str, source_definition_id: str):
+        self.source = source
+        self.source_id = source_id
+        self.source_definition_id = source_definition_id
+
+    @classmethod
+    def from_api_json(cls, api_json: Dict[str, Any]):
+        return cls(
+            AirbyteSource(
+                api_json["name"],
+                api_json["sourceName"],
+                api_json["connectionConfiguration"],
+            ),
+            api_json["sourceId"],
+            "",
+        )
+
+
+class AirbyteDestination:
+    """
+    Represents a user-defined Airbyte destination.
+    """
+
+    def __init__(self, name: str, destination_type: str, destination_configuration: Dict[str, Any]):
+        self.name = name
+        self.destination_type = destination_type
+        self.destination_configuration = destination_configuration
+
+    def must_be_recreated(self, other: "AirbyteDestination") -> bool:
+        return (
+            self.name != other.name
+            or self.destination_configuration != other.destination_configuration
+        )
+
+
+class InitializedAirbyteDestination:
+    """
+    User-defined Airbyte destination bound to actual created Airbyte destination.
+    """
+
+    def __init__(
+        self, destination: AirbyteDestination, destination_id: str, destination_definition_id: str
+    ):
+        self.destination = destination
+        self.destination_id = destination_id
+        self.destination_definition_id = destination_definition_id
+
+    @classmethod
+    def from_api_json(cls, api_json: Dict[str, Any]):
+        return cls(
+            AirbyteDestination(
+                api_json["name"],
+                api_json["destinationName"],
+                api_json["connectionConfiguration"],
+            ),
+            api_json["destinationId"],
+            "",
```

Comment:
in what cases is the destination definition id available? also, any reason to have this be an empty string rather than None?

Reply:
This isn't available when we pull from the API directly, we'd have to query it separately (& we don't end up using it in this case). Will change to `None` since it's more explicit.